packages/workspace/src/browser/workspace-service.ts

```diff
@@ -238,28 +274,34 @@ export class WorkspaceService implements FrontendApplicationContribution {
      * Adds a root folder to the workspace
      * @param uri URI of the root folder being added
      */
-    async addRoot(uri: URI): Promise<void> {
+    async addRoot(uri: URI, metaDataFile: URI | undefined): Promise<void> {
         await this.roots;
-        if (!this.opened || !this._workspaceFolder) {
-            throw new Error('Folder cannot be added as there is no active folder in the current workspace.');
+
+        if (metaDataFile) {
+            const uriStr = metaDataFile.withoutScheme().toString();
+            if (!await this.fileSystem.exists(uriStr)) {
+                await this.fileSystem.createFile(uriStr);
+            }
+            if (!this._workspaceMetaDataFile) {
+                this._workspaceMetaDataFile = await this.toFileStat(metaDataFile);
+            }
         }
 
-        const rootToAdd = uri.toString();
-        const valid = await this.toValidRoot(rootToAdd);
+        if (!this.opened) {
+            throw new Error('Folder cannot be added as there is no active workspace or opened folder.');
+        }
+        const valid = await this.toValidRoot(uri);
         if (!valid) {
-            throw new Error(`Invalid workspace root URI. Expected an existing directory location. URI: ${rootToAdd}.`);
+            throw new Error(`Invalid workspace root URI. Expected an existing directory location. URI: ${uri.toString()}.`);
         }
-        if (this._workspaceFolder && !this._roots.find(r => r.uri === valid.uri)) {
-            const configUri = this.getWorkspaceConfigFileUri();
-            if (configUri) {
-                if (!await this.fileSystem.exists(configUri.toString())) {
-                    await this.fileSystem.createFile(configUri.toString());
-                }
-                await this.writeRootFolderConfigFile(
-                    (await this.fileSystem.getFileStat(configUri.toString()))!,
-                    [...this._roots, valid]
-                );
+
+        if (this._workspaceMetaDataFile && !this._roots.find(r => r.uri === valid.uri)) {
+            if (!this.isBeingWatched(this._workspaceMetaDataFile.uri)) {
+                await this.startWatch(this._workspaceMetaDataFile);
+                this.updateTitle();
```

Comment:
the data flow gets complicated: I would have one method like `setWorkspaceFile` which will check whether a workspace file is changed and if does call accordingly `updatedTitle`, unwatch old file, watch a new file. Instead of calling these methods from here and there.